File: workout-curation/app/services/hermes.py
"""OpenRouter Hermes 추론 서비스."""

import json
import logging

import httpx

from app.config import settings
from app.services.prompts import MISSION_SYSTEM_PROMPT, get_active_prompt

logger = logging.getLogger(__name__)

MAX_TOKENS = {
    "recommendation": 800,
    "mission_text": 200,
    "encouragement": 100,
}

_HEADERS = {
    "Authorization": f"Bearer {settings.openrouter_api_key}",
    "Content-Type": "application/json",
    "HTTP-Referer": "https://workout-curation.app",
}


async def _call(
    system_prompt: str,
    user_prompt: str,
    model: str,
    max_tokens: int,
) -> dict:
    payload = {
        "model": model,
        "messages": [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt},
        ],
        "max_tokens": max_tokens,
        "temperature": 0.3,
        "response_format": {"type": "json_object"},
    }
    async with httpx.AsyncClient(timeout=30) as client:
        response = await client.post(
            settings.openrouter_base_url,
            json=payload,
            headers=_HEADERS,
        )
        response.raise_for_status()
        return response.json()


def _parse_recommendation(raw: dict) -> dict:
    """OpenRouter 응답에서 추천 JSON 추출 및 검증."""
    try:
        content = raw["choices"][0]["message"]["content"]
        parsed = json.loads(content)
        assert "recommendations" in parsed, "recommendations 키 누락"
        assert "top_pick" in parsed, "top_pick 키 누락"
        assert len(parsed["recommendations"]) >= 1, "추천 결과 없음"
        return parsed

    except (KeyError, IndexError) as e:
        logger.error("OpenRouter 응답 구조 오류: %s", e)
        raise ValueError("Hermes API 응답 형식 오류") from e

    except json.JSONDecodeError as e:
        logger.error("JSON 파싱 실패: %s", e)
        content = raw["choices"][0]["message"]["content"]
        start, end = content.find("{"), content.rfind("}") + 1
        if start != -1 and end > start:
            return json.loads(content[start:end])
        raise ValueError("JSON 파싱 불가") from e

    except AssertionError as e:
        logger.error("응답 유효성 검사 실패: %s", e)
        raise ValueError(str(e)) from e


async def get_sport_recommendations(user_prompt: str) -> dict:
    """설문 + RAG 컨텍스트 → Hermes 70B → 종목 추천."""
    prompt_cfg = get_active_prompt()
    try:
        raw = await _call(
            system_prompt=prompt_cfg["system"],
            user_prompt=user_prompt,
            model=settings.hermes_heavy,
            max_tokens=prompt_cfg["max_tokens"],
        )
        return _parse_recommendation(raw)
    except httpx.HTTPStatusError as e:
        if e.response.status_code == 429:
            raise RuntimeError("OpenRouter 요청 한도 초과") from e
        raise RuntimeError(f"API 오류 {e.response.status_code}") from e
    except httpx.TimeoutException:
        raise RuntimeError("Hermes API 응답 타임아웃 (30s)")


async def generate_mission_text(sport: str, facility_name: str, user_name: str) -> str:
    """미션 문장 생성 — Hermes 8B (비용 절약)."""
    user_prompt = f"{user_name}님을 위한 미션: {facility_name}에서 {sport} 체험. 친근한 한국어 1문장으로."
    try:
        raw = await _call(
            system_prompt=MISSION_SYSTEM_PROMPT,
            user_prompt=user_prompt,
            model=settings.hermes_light,
            max_tokens=MAX_TOKENS["mission_text"],
        )
        return raw["choices"][0]["message"]["content"].strip()
    except httpx.TimeoutException:
        return f"{facility_name}에서 {sport} 첫 체험에 도전해보세요!"

File: workout-frontend/lib/services/hermes.ts
const BASE_URL = 'https://openrouter.ai/api/v1/chat/completions';
const HEADERS = {
  'Authorization': `Bearer ${process.env.OPENROUTER_API_KEY}`,
  'Content-Type': 'application/json',
  'HTTP-Referer': 'https://workout-curation.app',
};

const HEAVY = process.env.HERMES_HEAVY ?? 'nousresearch/hermes-3-llama-3.1-70b';
const LIGHT = process.env.HERMES_LIGHT ?? 'nousresearch/hermes-3-llama-3.1-8b';

const RECOMMENDATION_SYSTEM_PROMPT = `당신은 운동 큐레이션 전문가입니다.
사용자의 성향, MBTI, 운동 이력, 라이프스타일을 분석하여 최적의 운동 종목을 추천합니다.
반드시 아래 JSON 형식으로만 응답하세요.

{
  "recommendations": [
    {
      "sport": "종목명",
      "reason": "이 사람에게 맞는 이유 (2-3문장, 목표·신체상태·이력 반영)",
      "difficulty": "입문자도 괜찮음 | 약간의 체력 필요 | 도전적",
      "first_mission": "이번 주말 할 수 있는 구체적인 미션 (1문장)"
    }
  ],
  "top_pick": "종목명",
  "encouragement": "사용자에게 전하는 한 마디 (1문장)"
}`;

export const CHAT_SYSTEM_PROMPT = `당신은 친절한 운동 추천 AI 챗봇입니다.
사용자와 자연스러운 한국어 대화를 통해 아래 순서대로 정보를 수집하세요.

━━━ 수집 순서 ━━━

[1단계] 기본 정보
 - 이름, 나이

[2단계] 운동 목적 (가장 중요)
 - "운동을 통해 가장 얻고 싶은 게 뭐예요?" 라고 물어보세요
 - 보기: 스트레스 해소 / 체중 감량 / 체력·건강 유지 / 새로운 사람 만나기 / 성취감·도전

[3단계] 신체 상태
 - 불편한 관절·부위가 있는지 (무릎/허리/어깨/없음)
 - 현재 체력 수준 (낮음: 계단 오르면 숨참 / 보통: 30분 걷기 가능 / 양호: 가끔 운동 가능)

[4단계] 시간·예산
 - 운동 가능 시간대 (평일 아침 / 평일 저녁 / 주말)
 - 한 번에 쓸 수 있는 시간 (30분 이내 / 1시간 / 1시간 이상)
 - 월 운동 예산 (3만원 이하 / 5~10만원 / 10만원 이상)

[5단계] 환경·분위기 선호
 - 실내 vs 실외 선호
 - 혼자 / 1~2명과 함께 / 여러 사람과 활기차게

[6단계] 운동 경험
 - 이전에 운동한 적 있는지
 - 있다면: 어떤 운동이었는지, 좋았던 점과 그만둔 이유
 - 없다면 이 질문 건너뜀

[7단계] MBTI (선택)
 - "혹시 MBTI 아세요? 모르셔도 괜찮아요!" 라고 가볍게 물어보세요

━━━ 대화 규칙 ━━━
- 한 번에 하나의 질문만 하세요
- 친근하고 따뜻한 말투 유지
- 사용자 답변에 짧게 공감한 뒤 다음 질문으로 자연스럽게 넘어가세요
- 모든 단계 완료 시 survey_complete: true 설정

━━━ 응답 형식 (반드시 JSON만 출력) ━━━

수집 중:
{"reply": "응답 텍스트", "survey_complete": false}

완료 시:
{
  "reply": "완벽해요! 지금까지 알려주신 내용으로 딱 맞는 운동을 찾아드릴게요 🔍",
  "survey_complete": true,
  "survey_data": {
    "user_name": "이름",
    "age": 나이(숫자),
    "gender": "남성|여성|기타|미응답",
    "mbti": "MBTI 또는 모름",
    "goal": "스트레스해소|체중감량|체력유지|사교|성취감",
    "physical_limit": "무릎|허리|어깨|없음",
    "fitness_level": "낮음|보통|양호",
    "session_duration": "30분이내|1시간|1시간이상",
    "environment": "실내|실외|상관없음",
    "had_exercise": true,
    "past_sport": "이전 운동 종목 또는 null",
    "liked_aspect": "좋았던 점 또는 null",
    "quit_reason": "그만둔 이유 또는 null",
    "activity_level": "거의 없음|주 1-2회|주 3회 이상",
    "preferred_time": "아침|저녁|주말",
    "social_pref": "혼자|소수|단체",
    "stress_style": "격렬하게|조용하게|창의적으로",
    "budget": 50000,
    "experience": "경험 요약 1~2문장",
    "avoid": "신체 제약 기반 기피 요소"
  }
}`;

async function callOpenRouter(
  messages: { role: string; content: string }[],
  model: string,
  maxTokens: number,
): Promise<string> {
  const res = await fetch(BASE_URL, {
    method: 'POST',
    headers: HEADERS,
    body: JSON.stringify({
      model,
      messages,
      max_tokens: maxTokens,
      temperature: 0.3,
      response_format: { type: 'json_object' },
    }),
  });
  if (!res.ok) throw new Error(`OpenRouter error: ${res.status}`);
  const data = await res.json();
  return data.choices[0].message.content as string;
}

export async function getSportRecommendations(userPrompt: string): Promise<{
  recommendations: { sport: string; reason: string; difficulty: string; first_mission: string }[];
  top_pick: string;
  encouragement: string;
}> {
  const content = await callOpenRouter(
    [
      { role: 'system', content: RECOMMENDATION_SYSTEM_PROMPT },
      { role: 'user', content: userPrompt },
    ],
    HEAVY,
    800,
  );
  return JSON.parse(content);
}

export async function generateMissionText(
  sport: string,
  level: number,
  userName: string,
): Promise<string> {
  const prompt = `운동 종목: ${sport}, 레벨: ${level}, 사용자 이름: ${userName}
레벨별 미션:
- 레벨1: 체험권/1일권으로 첫 방문하는 미션
- 레벨2: 2-3회 방문해서 기본기를 배우는 미션
- 레벨3: 월 정기권 or 동호회 가입하는 미션
위 레벨에 맞는 구체적인 미션 문장 1개만 JSON으로 반환: {"mission": "미션 문장"}`;

  const content = await callOpenRouter(
    [{ role: 'user', content: prompt }],
    LIGHT,
    200,
  );
  const parsed = JSON.parse(content);
  return parsed.mission as string;
}

export async function chatWithHermes(
  messages: { role: string; content: string }[],
): Promise<{ reply: string; survey_complete: boolean; survey_data?: Record<string, unknown> }> {
  const res = await fetch(BASE_URL, {
    method: 'POST',
    headers: HEADERS,
    body: JSON.stringify({
      model: LIGHT,
      messages,
      max_tokens: 600,
      temperature: 0.5,
      response_format: { type: 'json_object' },
    }),
  });
  if (!res.ok) throw new Error(`OpenRouter error: ${res.status}`);
  const data = await res.json();
  const content = data.choices[0].message.content as string;
  const parsed = JSON.parse(content);
  return {
    reply: parsed.reply ?? '...',
    survey_complete: parsed.survey_complete ?? false,
    survey_data: parsed.survey_data,
  };
}
